### PR TITLE
Vmi mmap guest 2

### PR DESCRIPTION
**Pull request by rageagainsthepc**

### CODE REVIEW

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -62,6 +62,9 @@ target_link_libraries(vmi-dump-memory vmi_shared)
 add_executable(vmi-module-list module-list.c)
 target_link_libraries(vmi-module-list vmi_shared)
 
+add_executable(vmi-memory-mapping vmi-memory-mapping.c)
```
Comment (thomasdangl) [inline examples/CMakeLists.txt:65]: Not sure if this should be part of the PR? If so, it should be part of the ENABLE_WINDOWS block below

File: libvmi/libvmi.h [hunk 2]
```diff
@@ -1382,6 +1396,39 @@ status_t vmi_mmap_guest(
...
+ * physical address are skipped.
+ *
+ * @param[in] vmi LibVMI instance
+ * @param[in] ctx Access context
+ * @param[in] num_pages Number of guest pages to be mapped (starting from ctx.addr)
+ * @param[out] mapped_regions
+ * @return VMI_SUCCESS if mapping succeeds, VMI_FAILURE otherwise
+ */
+status_t
+vmi_mmap_guest_2(
```
Comment (thomasdangl) [inline libvmi/libvmi.h:1418]: Maybe we could come up with a speaking name?

File: libvmi/read.c [hunk 3]
```diff
@@ -145,9 +145,187 @@ vmi_mmap_guest(
...
+        vmi_free_mapped_regions(result);
+    }
+
     return ret;
 }
 
+void vmi_free_mapped_regions(mapped_regions_t *mapping)
+{
+    if (mapping) {
+        for(size_t i = 0; i < mapping->size; i++) {
```
Comment (thomasdangl) [inline libvmi/read.c:322]: Formatting

File: libvmi/read.c [hunk 4]
```diff
@@ -145,9 +145,187 @@ vmi_mmap_guest(
...
+        // resize array if necessary
+        if (result->size >= result_capacity) {
+            result_capacity *= 2;
+            void *temp = realloc((void *) result, sizeof(mapped_regions_t) + sizeof(mapped_region_t) * result_capacity);
+
+            if (!temp) {
+                goto done;
+            }
+
+            result = (mapped_regions_t*) temp;
```
Comment (thomasdangl) [inline libvmi/read.c:284]: Why are you casting here?

File: libvmi/read.c [hunk 5]
```diff
@@ -145,9 +145,187 @@ vmi_mmap_guest(
...
+        }
+
+        // add current memory region
+        result->regions[result->size++] = cur_region;
+    }
+
+    // reduce the capacity of the array to the actual number of elements in order to not waste memory
+    void *temp = realloc((void *) result, sizeof(mapped_regions_t) + sizeof(mapped_region_t) * result->size);
```
Comment (thomasdangl) [inline libvmi/read.c:292]: Here as well

File: libvmi/read.c [hunk 6]
```diff
@@ -145,9 +145,187 @@ vmi_mmap_guest(
...
+
+    if (!temp) {
+        goto done;
+    }
+
+    result = (mapped_regions_t*) temp;
```
Comment (thomasdangl) [inline libvmi/read.c:298]: And here?

File: examples/vmi-memory-mapping.c [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        goto cleanup;
+    }
+
+    access_context_t accessContext = {0};
+    accessContext.version = ACCESS_CONTEXT_VERSION;
+    accessContext.translate_mechanism = VMI_TM_PROCESS_PID;
+    accessContext.pid = 0;
+    accessContext.addr = list_head;
+
+    if (vmi_mmap_guest_2(vmi, &accessContext, 1000, &mapping) == VMI_FAILURE)
```
Comment (thomasdangl) [inline examples/vmi-memory-mapping.c:55]: If this is to be comitted, clean it up around here ...

File: examples/CMakeLists.txt (hunk 1 above)
Comment (rageagainsthepc) [reply examples/CMakeLists.txt:65]: Including an example should make testing easier and therefore increase our chances of having this merged quickly.

File: libvmi/libvmi.h (hunk 2 above)
Comment (rageagainsthepc) [reply libvmi/libvmi.h:1418]: By all means, go ahead ;)

File: libvmi/libvmi.h [hunk 8]
```diff
@@ -722,6 +722,20 @@ typedef struct page_info {
...
+ */
+typedef struct mapped_region {
+    addr_t start_va;
+    size_t num_pages;
+    void** access_ptrs;
+} mapped_region_t;
+
+typedef struct mapped_regions {
+    size_t size;
+    mapped_region_t regions[0];
```
[the remark this answers is not in the document]
Comment (thomasdangl) [reply libvmi/libvmi.h:736]: No, otherwise I would not suggest the change 🤣

**Pull request closed without merging**